internal/datastore/spanner/watch.go

```diff
@@ -53,216 +53,76 @@ func parseDatabaseName(db string) (project, instance, database string, err error
-		var context map[string]any
-		if contextString != nil {
-			if err := json.Unmarshal([]byte(contextString.(string)), &context); err != nil {
-				return nil, err
-			}
+	sendError := func(err error) {
+		if errors.Is(ctx.Err(), context.Canceled) {
+			errs <- datastore.NewWatchCanceledErr()
+			return
 		}
 
-		return common.ContextualizedCaveatFrom(name, context)
+		errs <- err
 	}
-	return nil, nil
-}
-
-const maxSchemaWatchRetryCount = 15
-
-func (sd spannerDatastore) WatchSchema(ctx context.Context, afterRevision datastore.Revision) (<-chan *datastore.SchemaState, <-chan error) {
-	updates := make(chan *datastore.SchemaState, 10)
-	errs := make(chan error, 1)
-
-	lastRevision := afterRevision
-	retryCount := -1
-	go func() {
-		running := true
-		for running {
-			retryCount++
-			retryHistogram.Observe(float64(retryCount))
-			if retryCount >= maxSchemaWatchRetryCount {
-				errs <- fmt.Errorf("schema watch terminated after %d retries", retryCount)
-				return
-			}
-
-			log.Debug().Str("revision", lastRevision.String()).Int("retry-count", retryCount).Msg("starting schema watch connection")
-
-			sd.watchSchemaWithoutRetry(ctx, lastRevision, func(update *datastore.SchemaState, err error) {
-				if err != nil {
-					// TODO(jschorr): If this is a terminal error, just quit instead of retrying.
-					running = true
-
-					// Sleep a bit for retrying.
-					pgxcommon.SleepOnErr(ctx, err, uint8(retryCount))
-					return
-				}
 
-				if update.Revision.GreaterThan(lastRevision) {
-					lastRevision = update.Revision
-				}
+	sendChange := func(change *datastore.RevisionChanges) bool {
+		select {
+		case updates <- change:
+			return true
 
-				updates <- update
-				retryCount = 0
-			})
+		default:
+			return false
 		}
-	}()
-
-	return updates, errs
-}
+	}
 
-func (sd spannerDatastore) watchSchemaWithoutRetry(ctx context.Context, afterRevisionRaw datastore.Revision, processUpdate func(update *datastore.SchemaState, err error)) {
 	project, instance, database, err := parseDatabaseName(sd.database)
 	if err != nil {
-		processUpdate(nil, err)
+		sendError(err)
 		return
 	}
 
+	// Select the change stream to use for the watch.
+	// TODO(jschorr): we can probably just get rid of the non-combined stream, given the filter below.
+	changeStreamName := CombinedChangeStreamName
+	if opts.Content&datastore.WatchRelationships == opts.Content {
+		changeStreamName = RelationTupleChangeStreamName
+	}
+	if opts.Content&(datastore.WatchSchema|datastore.WatchCheckpoints) == opts.Content {
+		changeStreamName = SchemaChangeStreamName
+	}
+
 	afterRevision := afterRevisionRaw.(revision.Decimal)
 	reader, err := changestreams.NewReaderWithConfig(
 		ctx,
 		project,
 		instance,
 		database,
-		SchemaChangeStreamName,
+		changeStreamName,
 		changestreams.Config{
 			StartTimestamp:    timestampFromRevision(afterRevision),
-			HeartbeatInterval: sd.config.schemaWatchHeartbeat,
+			HeartbeatInterval: heartbeatInterval,
```

Comment:
>  I wonder how that never surfaced as a problematic.

Likely because no one is using it

> In any case, we are now moving to a 100 Millisecond (!!!) default interval. Folks out there using SpiceDB Watch API and Spanner, who upgrade to this next release, will see a sudden increase of CPU usage on their Spanner that is likely to cause knock on effect unless their are overprovisioned.

I'll change it to 1 second then